src/porepy/numerics/time_step_control.py

```diff
@@ -1,130 +1,191 @@
-            s += " upper optimal iteration range."
-            raise ValueError(s)
-        elif iter_optimal_range[1] > iter_max:
-            s = "Upper optimal iteration range cannot be larger than"
-            s += " maximum number of iterations."
-            raise ValueError(s)
-        elif iter_optimal_range[0] < 0:
-            s = "Lower optimal iteration range cannot be negative."
-            raise ValueError(s)
-
-        # Checks for lower and upper multiplication factors
-        if iter_lowupp_factor[0] <= 1:
-            raise ValueError("Expected lower multiplication factor > 1.")
-        elif iter_lowupp_factor[1] >= 1:
-            raise ValueError("Expected upper multiplication factor < 1.")
-
-        # Checks for sensible combinations of iter_optimal_range and iter_lowupp_factor
-        if dt_min_max[0] * iter_lowupp_factor[0] > dt_min_max[1]:
-            raise ValueError("Encountered dt_min * iter_low_factor > dt_max.")
-        elif dt_min_max[1] * iter_lowupp_factor[1] < dt_min_max[0]:
-            raise ValueError("Encountered dt_max * iter_upp_factor < dt_min.")
-
-        # Check for recomputation factor
-        if recomp_factor >= 1:
-            raise ValueError("Expected recomputation multiplication factor < 1.")
-
-        # Check for maximum number of recomputation attempts
-        if recomp_max <= 0:
-            raise ValueError("Number of recomputation attempts must be > 0.")
+        # Set dt_min_max if necessary
+        if dt_min_max is None:
+            dt_min_max = (0.001 * schedule[-1], 0.1 * schedule[-1])
+
+        # More sanity checks below. Note that all the remaining sanity checks (but one) are
+        # only needed when constant_dt = False. Thus, to save time when constant_dt = True,
+        # we use this rather ugly if-statement from below.
+
+        if not constant_dt:
+
+            # Sanity checks for dt_min and dt_max
+            if dt_init < dt_min_max[0]:
+                raise ValueError(
+                    "Initial time step cannot be smaller than minimum time step."
+                )
+            elif dt_init > dt_min_max[1]:
+                raise ValueError(
+                    "Initial time step cannot be larger than maximum time step."
+                )
+
+            # NOTE: The above checks guarantee that minimum time step <= maximum time step
+
+            # Sanity checks for maximum number of iterations.
+            # Note that 0 is a possibility. This will imply that the solver reaches
+            # convergence directly, e.g., as in linear solvers.
+            if iter_max < 0:
+                raise ValueError("Maximum number of iterations must be non-negative.")
+
+            # Sanity checks for optimal iteration range
+            if iter_optimal_range[0] > iter_optimal_range[1]:
+                s = "Lower optimal iteration range cannot be larger than"
+                s += " upper optimal iteration range."
+                raise ValueError(s)
```

Comment:
In such cases, I like it when the error message prints out the values given, it makes debugging easier, but I'm never sure if it is worth the added code complexity (small, yes, but so is the debugging effort). Any opinions?

Reply:
As you mentioned, the only drawback here is more code complexity + slightly less readable code. I'm in favour of adding the actual value in the error message in these cases, considering the number of parameters that the user has to deal with.

Reply:
I agree. The code is not really more complex in this case using string formatting.

Reply:
Ok. I will add the actual values to the error messages.